### PR TITLE
Add paste service utility

### PR DESCRIPTION
**Pull request by ChrisLovering**
Docs:
image

changelog:
image

### CODE REVIEW

File: pydis_core/utils/paste_service.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+from aiohttp import ClientConnectorError, ClientSession
+
+from pydis_core.utils import logging
+
+log = logging.get_logger(__name__)
+
+FAILED_REQUEST_ATTEMPTS = 3
+MAX_PASTE_SIZE = 256 * 1024  # 256kB
```
Comment (mbaruh) [inline pydis_core/utils/paste_service.py:8]: Not critical, but I'd go for 128. Closer to the old limit and still a power of 2 :D

File: pydis_core/utils/paste_service.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    """Raised when an error is encountered uploading to the paste service."""
+
+
+class PasteTooLongError(Exception):
+    """Raised when content is too large to upload to the paste service."""
+
+
+async def send_to_paste_service(
+    *,
+    file_name: str,
```
Comment (mbaruh) [inline pydis_core/utils/paste_service.py:21]: The name field is optional in the API, so I don't think we need to require it.

File: pydis_core/utils/paste_service.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            continue
+
+        if response.status == 400:
+            log.warning(
+                f"Paste service returned error {response_json['message']} with status code {response.status}, "
+                f"trying again ({attempt}/{FAILED_REQUEST_ATTEMPTS})."
+            )
+            continue
+
+        if link := "link" in response_json:
```
Comment (mbaruh) [inline pydis_core/utils/paste_service.py:86]: I don't understand this line, how does it form the paste link? Seems like `link` is a boolean.

**Review by mbaruh — CHANGES_REQUESTED** [review]

File: pydis_core/utils/paste_service.py (hunk 1 above)
Comment (ChrisLovering) [reply pydis_core/utils/paste_service.py]: Updated in the [force push](https://github.com/python-discord/bot-core/compare/506329a56f2642af5ff7eed53a5ced1838fe4336..801e1e5151cde2d01024fc025b8b688558720070)

File: pydis_core/utils/paste_service.py (hunk 2 above)
Comment (ChrisLovering) [reply pydis_core/utils/paste_service.py]: Updated in the [force push](https://github.com/python-discord/bot-core/compare/506329a56f2642af5ff7eed53a5ced1838fe4336..801e1e5151cde2d01024fc025b8b688558720070)

File: pydis_core/utils/paste_service.py (hunk 3 above)
Comment (ChrisLovering) [reply pydis_core/utils/paste_service.py]: hah yea, not sure what I was thinking here either. Updated in the [force push](https://github.com/python-discord/bot-core/compare/506329a56f2642af5ff7eed53a5ced1838fe4336..801e1e5151cde2d01024fc025b8b688558720070)

File: pydis_core/utils/paste_service.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            continue
+
+        if response.status == 400:
+            log.warning(
+                f"Paste service returned error {response_json['message']} with status code {response.status}, "
+                f"trying again ({attempt}/{FAILED_REQUEST_ATTEMPTS})."
+            )
+            continue
+
+        if link := response_json.get("link" ):
```
Comment (mbaruh) [inline pydis_core/utils/paste_service.py:86]:
```suggestion
        if link := response_json.get("link"):
```
Do we not have linting here?

**Review by mbaruh — APPROVED** [review]
A couple of things that might make sense to have but aren't blockers:
- Enable the "Use a longer URI" option, although I don't see it in the API docs.
- Return the deletion URL. This would allow us to cache it and potentially use it via a command if we find that a paste contains sensitive information.

File: pydis_core/utils/paste_service.py (hunk 4 above)
Comment (ChrisLovering) [reply pydis_core/utils/paste_service.py]: We _should_ not sure why this wasn't picked up. I've updated this now anyway, to simply check for a 200 status code.

**Comment by ChrisLovering** [comment]
> Enable the "Use a longer URI" option, although I don't see it in the API docs.
> Return the deletion URL. This would allow us to cache it and potentially use it via a command if we find that a paste contains sensitive information.

I like both of these, added in a force push.

File: pydis_core/utils/paste_service.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    """
+    Upload some contents to the paste service.
+
+    Args:
+        contents: The content to upload to the paste service.
+        paste_url: The base url to the paste service.
+        http_session (aiohttp.ClientSession): The session to use when POSTing the content to the paste service.
+        file_name: The name of the file to save to the paste service.
+        lexer: The lexer to save the content with.
+        max_size: The maximum number of bytes to be allowed. Anything larger that 256kB will be rejected.
```
Comment (vivekashok1221) [inline pydis_core/utils/paste_service.py:52]: This needs to be updated.
```suggestion
        max_size: The maximum number of bytes to be allowed. Anything larger that 128kB will be rejected.
```

**Review by vivekashok1221 — CHANGES_REQUESTED** [review]

File: pydis_core/utils/paste_service.py (hunk 5 above)
Comment (ChrisLovering) [reply pydis_core/utils/paste_service.py]: Updated in the [forcepush](https://github.com/python-discord/bot-core/compare/9478a6fcbda210b7b8ad4ee83bc8ba3e89ff5498..7345b10961af47ca179d8b7009c0d7d0eb3c08e2) to reference the constant instead. Looks like this now

image

image

**Review by vivekashok1221 — APPROVED** [review]
LGTM :star2:.

**Comment by vivekashok1221** [comment]
Good ol' reliable Chris.

**Pull request merged**